### ISSUE TITLE
Cortex M4F core::fmt f32 miss optimisation 

### ISSUE THREAD

**Issue by no111u3** [rust-lang/rust#118337]
Adding formatting output (`write`) for float numbers for Cortex M4F increase binary size twice: from 39396 bytes to 66468 bytes.
The main high weigh functions are (provided by `cargo bloat`):
 ```
 2.4%  22.7% 11.2KiB             std core::fmt::float::float_to_decimal_common_shortest
 1.9%  17.8%  8.8KiB             std core::fmt::float::float_to_decimal_common_exact
 0.9%   8.6%  4.2KiB             std core::num::dec2flt::<impl core::str::traits::FromStr for f32>::from_str
 0.4%   3.8%  1.9KiB             std core::fmt::Formatter::pad
 0.3%   2.9%  1.4KiB             std core::num::flt2dec::strategy::dragon::mul_pow10
 ```
and others

For desktop application (Apple M2Pro chip, native build) with same functions size lover than few kilobytes:
```
 0.1%   0.3%     732B         std core::fmt::float::float_to_decimal_common_exact
 0.1%   0.3%     624B         std core::fmt::float::float_to_decimal_common_shortest
  0.3%   0.6%   1.3KiB         std core::num::dec2flt::<impl core::str::traits::FromStr for f32>::from_str
 0.2%   0.4%     980B         std core::fmt::Formatter::pad
```

Which reason of overweigh these functions for Cortex-M4F?

**Comment by workingjubilee** [comment]
The two CPUs you mention do not share an instruction set, and there are characteristics of the latter instruction set that can make it more efficient at expressing certain code patterns. It's not clear whether you're using the AArch32 or Thumb32 ISA by your report, but neither of those are AArch64.

It is, of course, likely this is LLVM performing subpar optimization, but it would be best if you could provide significantly more information about your compilation settings so people are more able to diagnose and fix this.

**Comment by no111u3** [comment]
@workingjubilee [gist with std core::fmt::float::float_to_decimal_common_shortest](https://gist.github.com/no111u3/dcf580af4d3bf9e1bf79bb32db7f8bf6)

**Comment by workingjubilee** [comment]
@no111u3 I asked for the compilation settings you used, not the assembly text, and _not_ just the optimization level. There is no Cortex M4F target from Rust's perspective: there is only armv7m and thumbv7m. The Cortex M4F core can execute either instruction set, much the same as x86-64 CPUs can execute code both in and out of long mode, but some instructions are differently encoded. ( The belief that the two de facto x86 instruction sets are completely compatible is popular but incorrect. )

You also said the `-Ctarget-cpu=native` target was for an M2 device, but you did not specify Linux or Darwin. I think it would be reasonable to assume you're using the Apple OS on an Apple Silicon chip, but you didn't specify if that is also `-Copt-level=z`, and that is another detail.

Bugs on the Rust issue tracker stay open for sometimes very long times, and it is important that the instructions to reveal the bug are reproducible, even with a future compiler. It is common for missed optimization bugs to have been fixed but no one knows whether there is a bug still because they cannot reproduce the problem, _because it has been fixed_, and they don't know whether the instructions are complete.

**Comment by workingjubilee** [comment]
Checking the asm more closely, it has to be one of the thumb targets because of the usage of `IT` instructions. But _please_ explain what the actual `--target` and `-Ctarget-cpu` was, in terms of strings passed to the compiler. The compiler doesn't actually know what a CPU is, it only understands abstractions.

**Comment by no111u3** [comment]
@workingjubilee I'm not sure that arguments use for build `core` library but I use `thumbv7em-none-eabihf` for my project and these arguments for build:
```toml
[profile.release]
opt-level = 'z'
debug = false
codegen-units = 1
incremental = false
lto = true
```

**Comment by workingjubilee** [comment]
Ah, you are not using `-Zbuild-std`? That clarifies some things. My apologies. I had misimagined, based on your previous remarks, that you had rebuilt the standard library using `-Zbuild-std` as part of your efforts. If you are using the prebuilt standard library, the optimization level you set for your code won't really matter, except for code that gets generated on-site into your binary, and I do not believe the stdlib is optimized for code size, even for that target.

The lexical crate, however, _will_ get optimized according to `-Copt-level=z`, and thus it will focus on code size reduction.

I am examining the x86-64 version of this code and I see that LLVM is legalizing many float operations and predicated instructions (cmov and friends). I wonder how that is affected by Thumb's use of IT (if-then) for predication? But on disassembling the section in question for the Thumb target on the latest nightly, I see many strange pieces of disassembly, such as `movs r0, r0` appearing many times in a row, which suggests to me I should not trust the disassembly, either.

**Comment by workingjubilee** [comment]
Oh, this works. Perhaps it overrides the use of the filename for guiding disassembly? Not sure.

```shell
cat $RLIB | llvm-objdump --arch-name=thumb -D - >"$RLIB.asm"
```

**Comment by workingjubilee** [comment]
@no111u3 I conferred with other members of WG-binary-size and it seems like the disassembly doesn't matter, and I was slightly off too. Enabling LTO *should* mean that the stdlib's `.llvmbc` section is taken and used to freshly build the float formatting code into your binary with the optimization settings you set.

**Comment by thejpster** [comment]
An idle thought - is the code using f32 or f64? I suspect the M4F might only have a single precision FPU.

**Comment by nikic** [comment]
> Enabling LTO _should_ mean that the stdlib's `.llvmbc` section is taken and used to freshly build the float formatting code into your binary with the optimization settings you set.

Just a note on this, the opt-level for libstd still matters because it will be used for pre-link optimization. Only the post-link optimization will use the opt-level specified in the profile.

**Comment by nikic** [comment]
> I see! So it still matters then because code that is optimized for speed that does a bunch of unrolling then has to be minimized by the same optimizer, which may then have less possible "paths" to reduce the code size, as well as produced a bunch of work for itself to undo?

Yes. Inlining (within libstd only) and full unrolling happens pre-link, so those things will use `opt-level=3` heuristics, and these transforms cannot be undone. LTO inlining, runtime unrolling and vectorization will then use `opt-level=z` heuristics.

(I think ... I hope I'm not getting rustc's compilation model wrong here.)

**Comment by cbiffle** [comment]
Hi! For what it's worth, during the Hubris toolchain upgrade from something-ancient (late 2022) to current nightly, we've had a couple of tasks explode in size due to what I _think_ is the same behavior the OP is reporting. The floating point formatting code (which we're pulling in by accident, but no matter) has gotten significantly larger.

This is on a Cortex-M7 using `thumbv7em-none-eabihf`, so we have hardware support for both dual- and single-precision floating point.

These specific items seem to be contributing the bulk of the difference; they are new in the past year or two:

```
0000026e t core::num::flt2dec::strategy::dragon::mul_pow10
00000246 t core::num::bignum::Big32x40::mul_digits
00001668 t core::num::flt2dec::strategy::dragon::format_shortest
00000162 t core::num::bignum::Big32x40::mul_pow2
000008aa t core::num::flt2dec::strategy::grisu::format_shortest_opt
0000002c t core::panicking::panic_const::panic_const_div_by_zero
000001c6 t core::num::flt2dec::strategy::grisu::format_exact_opt::possibly_round
000000e0 t core::num::flt2dec::digits_to_dec_str
00000008 r core::num::flt2dec::strategy::dragon::POW10TO16
00000010 r core::num::flt2dec::strategy::dragon::POW10TO32
0000001c r core::num::flt2dec::strategy::dragon::POW10TO64
00000038 r core::num::flt2dec::strategy::dragon::POW10TO128
0000006c r core::num::flt2dec::strategy::dragon::POW10TO256
00000510 r core::num::flt2dec::strategy::grisu::CACHED_POW10
```
(columns are: size in hex, text or rodata, demangled symbol)
(In case you don't want to add hex numbers, that sums to 11,746 bytes.)

I'm guessing this code is making a time-memory tradeoff to accelerate float to decimal conversions; the tradeoff may make sense on large virtual-memory systems, and may not make sense on very small systems.

Hope that helps.

**Comment by thejpster** [comment]
Out of interest, do you pass additional target-features to enable the Double Precision support (`+fp64`) because I dimly recall that target only enables single precision support so as to provide the widest compatibility.

**Comment by cbiffle** [comment]
> Out of interest, do you pass additional target-features to enable the Double Precision support (`+fp64`) because I dimly recall that target only enables single precision support so as to provide the widest compatibility.

That's a fair point -- we all have the stm32f7 to thank for that silliness. We are not explicitly forcing double-precision on, in part because the feature is unstable and makes the build _really noisy._

However,

1. I've found the code responsible for pulling in all this stuff, and it is definitely `f32` only.
2. I've tested a build using fp64 and it doesn't seem to affect the size of this code. The overall size of this binary changes by only 8 bytes.
